scylla-cdc-base/src/main/java/com/scylladb/cdc/cql/driver3/Driver3MasterCQL.java

```diff
@@ -20,40 +23,176 @@
+            // If we found a "rewritten" row, that means that
+            // we can use the "Streams description table V2".
+            logger.atFiner().log("Using new (V2) streams description table, because a 'rewritten' row was found previously.");
+            return CompletableFuture.completedFuture(false);
+        }
+
+        // We haven't seen a rewritten row yet. Do a query
+        // to check if it exists now.
+
+        return executeOne(getFetchRewritten()).thenApply(fetchedRewritten -> {
+            if (fetchedRewritten.isPresent()) {
+                // There is a "rewritten" row.
+                foundRewritten = true;
+                logger.atInfo().log("Found a 'rewritten' row. Will use new (V2) streams description table from now on.");
+                return false;
+            } else {
+                logger.atWarning().log("Using legacy (V1) streams description table, even though newer (V2) table was found, but " +
+                        "a 'rewritten' row is still missing. This might mean that the rewriting process is still pending or you have " +
+                        "disabled streams description rewriting - in that case the library will not switch to the new (V2) table " +
+                        "until it discovers a 'rewritten' row. Read more at: " +
+                        "https://github.com/scylladb/scylla/blob/master/docs/design-notes/cdc.md#streams-description-table-v1-and-rewriting");
+                return true;
+            }
+        });
     }
 
-    private static RegularStatement getFetchStreams() {
-        return select().column("streams").from("system_distributed", "cdc_streams_descriptions")
-                .where(eq("time", bindMarker())).allowFiltering();
+    private CompletableFuture<PreparedStatement> getLegacyFetchSmallestGenerationAfter() {
+        if (legacyFetchSmallestGenerationAfterStmt != null) {
+            return CompletableFuture.completedFuture(legacyFetchSmallestGenerationAfterStmt);
+        } else {
+            ListenableFuture<PreparedStatement> prepareStatement = session.prepareAsync(
+                    select().min(column("time")).from("system_distributed", "cdc_streams_descriptions")
+                            .where(gt("time", bindMarker())).allowFiltering()
+            );
+            return FutureUtils.convert(prepareStatement).thenApply(preparedStatement -> {
+                legacyFetchSmallestGenerationAfterStmt = preparedStatement;
+                return preparedStatement;
+            });
+        }
+    }
+
+    private CompletableFuture<PreparedStatement> getFetchSmallestGenerationAfter() {
+        if (fetchSmallestGenerationAfterStmt != null) {
+            return CompletableFuture.completedFuture(fetchSmallestGenerationAfterStmt);
+        } else {
+            ListenableFuture<PreparedStatement> prepareStatement = session.prepareAsync(
+                    select().min(column("time")).from("system_distributed", "cdc_generation_timestamps")
```

Comment:
Why do we need allow filtering here? Aren't all the timestamps in a single partition now? Couldn't we just query that single partition? @kbr- ?

Reply:
In fact, it seems that it is already a query to a single partition.

Reply:
cdc_generation_timestamps is ordered descending (new generations first).

Doing a query: `SELECT * FROM cdc_generation_timestamps WHERE time > last_generation LIMIT 1;` would return the latest generation, not the one closest to `last_generation`.

Probably could do a `ORDER BY time ASC` clause - will it be much better? (I vaguely remember Scylla having problems with reverse ordering - is that true?)

Reply:
But is allow filtering needed for queries inside a single partition?

Reply:
No.

So should I go for (`allowFiltering()` removed):
```java
select().min(column("time")).from("system_distributed", "cdc_generation_timestamps")
   .where(eq("key", "timestamps")).and(gt("time", bindMarker()))
```

or (without `min`):
```java
select().column("time").from("system_distributed", "cdc_generation_timestamps")
        .where(eq("key", "timestamps")).and(gt("time", bindMarker()))
        .orderBy(asc("time")).limit(1)
```
?

I have already tested the second solution on 3-generation cluster (data inserted in each generation, it correctly went through all generations).

Reply:
I think just removing allow filtering is enough. Limit helps little because the coordinator still has to receive more data to make sure we're able to return something when tombstones are present on other replicas.

Reply:
Pushed amended commit with `allowFiltering()` removed.